### PR TITLE
[api] Refactor most of `Decl_kinds`

### PR DESCRIPTION
**Pull request by ejgallego**
We move the bulk of `Decl_kinds` to a better place [namely
`interp/decls`] and refactor the use of this information quite a bit.

The information seems to be used almost only for `Dumpglob`, so it
certainly should end there to achieve a cleaner core.

In particular the use of "variable" data stored in Decls seems very suspicious.

IMO this may need a bit more work, but this should be a good start, I wonder what people who know more about this `variable data` stuff can say.

Overlays:
- https://github.com/LPCIC/coq-elpi/pull/69
- https://github.com/coq-community/paramcoq/pull/38
- https://github.com/mattam82/Coq-Equations/pull/226
- https://github.com/Mtac2/Mtac2/pull/218
- https://github.com/QuickChick/QuickChick/pull/172

### CODE REVIEW

**Comment by ejgallego** [comment]
Sadly this had to depend on the previous kind cleanup, last 3 commits are the relevant ones.

**Comment by ejgallego** [comment]
cc: #10417

**Comment by ejgallego** [comment]
Ok, this may need some thinking but IMHO the bulk is ready for review.

**Review by herbelin — COMMENTED** [review]
Trusting @ejgallego whose work on how to organize files I regularly appreciate.

**Comment by ejgallego** [comment]
I guess this can still be improved, however the PR feels like mature so the moment to get an assignee has IMHO arrived.

**Comment by ejgallego** [comment]
Failure is spurious.

File: tactics/hints.ml [hunk 1]
```diff
@@ -1315,12 +1315,16 @@ let project_hint ~poly pri l2r r =
   let c = Reductionops.whd_beta sigma (mkApp (c, Context.Rel.to_extended_vect mkRel 0 sign)) in
   let c = it_mkLambda_or_LetIn
     (mkApp (p,[|mkArrow a Sorts.Relevant (lift 1 b);mkArrow b Sorts.Relevant (lift 1 a);c|])) sign in
-  let id =
+  let name =
```
Comment (SkySkimmer) [inline tactics/hints.ml]: Strictly speaking this is indeed an id and it's the declare API using the wrong label. It doesn't matter much I guess.

File: tactics/declare.ml [hunk 2]
```diff
@@ -11,20 +11,23 @@
...
+open Safe_typing
 open Libobject
 open Lib
-open Impargs
-open Safe_typing
-open Cooking
-open Decl_kinds
+
+(* This was meant as error for schemes, not clear how it works *)
+exception SchemeAlreadyDeclared of Pp.t
```
Comment (SkySkimmer) [inline tactics/declare.ml]: >This was meant as error for schemes

Isn't it the error for when you do 2 definitions with the same name? It doesn't seem to have much to do with schemes.

**Review by SkySkimmer — COMMENTED** [review]
I'm confused by the AlreadyDeclared name change, see comment.

File: interp/decls.mli [hunk 3]
```diff
@@ -10,32 +10,77 @@
...
+(* Only used in dumpglob *)
 val variable_secpath : variable -> qualid
 val variable_kind : variable -> logical_kind
+
+(* User in Lemma, Very dubious *)
 val variable_opacity : variable -> bool
+
+(* Used in declare, very dubious *)
 val variable_context : variable -> Univ.ContextSet.t
 val variable_polymorphic : variable -> bool
```
Comment (SkySkimmer) [inline interp/decls.mli]: These 2 can be removed: #10461.

File: interp/decls.mli [hunk 4]
```diff
@@ -10,32 +10,77 @@
...
+  ; opaque:bool
+  ; univs:Univ.ContextSet.t
+  ; poly:bool
+  ; kind:logical_kind
+  }
 
 val add_variable_data : variable -> variable_data -> unit
+
+(* Not used *)
 val variable_path : variable -> DirPath.t
```
Comment (SkySkimmer) [inline interp/decls.mli]: Why keep it then?

File: pretyping/heads.ml [hunk 5]
```diff
@@ -41,8 +41,7 @@ let rec compute_head env = function
      | Some c -> kind_of_head env c)
   | EvalVarRef id ->
     (match lookup_named id env with
-     | LocalDef (_,c,_) when not (Decls.variable_opacity id) ->
-       kind_of_head env c
+     | LocalDef (_,c,_) -> kind_of_head env c
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply pretyping/heads.ml]: Thanks for launching the bench @SkySkimmer , funnily enough the top user of opaque `Lets` (odd-order) seems to be faster here.

File: tactics/hints.ml (hunk 1 above)
Comment (ejgallego) [reply tactics/hints.ml]: I had the same thought actually, indeed that's an inconsistency but then we usually speak of "the name of a constant" right? So seems to me that `name` may be a better label. I dunno, no problem in going back to `id`, but it seems more like a label for a UID / GUIID, etc....

File: tactics/declare.ml (hunk 2 above)
Comment (ejgallego) [reply tactics/declare.ml]: Indeed, I've amended that; I wonder why I did that change, I was digging in history and likely got lost.

File: interp/decls.mli (hunk 3 above)
Comment (ejgallego) [reply interp/decls.mli]: cool!

File: interp/decls.mli (hunk 4 above)
Comment (ejgallego) [reply interp/decls.mli]: I though we could address the refactoring of this module in its own PR.

**Comment by ejgallego** [comment]
Wrong label for exception fixed.

**Comment by SkySkimmer** [comment]
Bench said: https://ci.inria.fr/coq/view/benchmarking/job/benchmark-part-of-the-branch/740/console
~~~
┌────────────────────────┬─────────────────────────┬─────────────────────────────────────────────┬─────────────────────────────────────────────┬───────────────────────────────┬────────────────────────────┐
│                        │      user time [s]      │                 CPU cycles                  │              CPU instructions               │     max resident mem [KB]     │         mem faults         │
│                        │                         │                                             │                                             │                               │                            │
│           package_name │     NEW     OLD PDIFF   │               NEW               OLD PDIFF   │               NEW               OLD PDIFF   │        NEW        OLD PDIFF   │     NEW     OLD    PDIFF   │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│ coq-mathcomp-odd-order │ 1442.50 1458.75 -1.11 % │     4019138244673     4065469797816 -1.14 % │     6892441543259     6892542837063 -0.00 % │    1260556    1261184 -0.05 % │      47      47    +0.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│               coq-hott │  336.69  339.86 -0.93 % │      916733980534      923761093517 -0.76 % │     1395810923662     1395881699780 -0.01 % │     476688     476760 -0.02 % │     461     246   +87.40 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│  coq-mathcomp-fingroup │   51.22   51.61 -0.76 % │      142900339278      143564018087 -0.46 % │      186634618068      186627699278 +0.00 % │     561848     561932 -0.01 % │       6       6    +0.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│               coq-corn │ 1451.26 1459.70 -0.58 % │     4048098364168     4070176691336 -0.54 % │     6014993897014     6016086079992 -0.02 % │     890852     890964 -0.01 % │       2      90   -97.78 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│       coq-math-classes │  216.26  217.33 -0.49 % │      603926547891      606216779064 -0.38 % │      767971490360      768009387247 -0.00 % │     514432     514728 -0.06 % │      38      21   +80.95 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│             coq-geocoq │ 1930.25 1939.29 -0.47 % │     5383889804479     5410473326192 -0.49 % │     7841829160819     7841906972372 -0.00 % │    1255740    1255052 +0.05 % │     197     321   -38.63 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│ coq-fiat-crypto-legacy │ 6285.61 6312.88 -0.43 % │    17495599522421    17565490583381 -0.40 % │    28549599928989    28541812067200 +0.03 % │    2384160    2382872 +0.05 % │    1241    1285    -3.42 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│       coq-fiat-parsers │  634.21  636.91 -0.42 % │     1775033664920     1779992713008 -0.28 % │     2689610457186     2689406053915 +0.01 % │    3224380    3232320 -0.25 % │     366     735   -50.20 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│          coq-fourcolor │ 2159.80 2168.60 -0.41 % │     6024004942988     6048210323515 -0.40 % │    11362187341614    11361292780277 +0.01 % │     927492     927232 +0.03 % │      14       0     +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│              coq-flocq │  459.60  461.38 -0.39 % │     1279308999366     1284256664082 -0.39 % │     1776323854898     1776450666727 -0.01 % │    1148336    1142520 +0.51 % │     335     130  +157.69 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│     coq-mathcomp-field │  230.76  231.58 -0.35 % │      642860814926      644919880708 -0.32 % │      919014944992      919089449538 -0.01 % │     750228     750076 +0.02 % │       0      74  -100.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│  coq-mathcomp-solvable │  189.12  189.72 -0.32 % │      526662207439      528062321307 -0.27 % │      704260592764      704072098767 +0.03 % │     784608     784692 -0.01 % │      11       8   +37.50 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│ coq-mathcomp-ssreflect │   40.05   40.16 -0.27 % │      110726732783      110760794607 -0.03 % │      133923775340      133873444937 +0.04 % │     520132     520276 -0.03 % │     139      10 +1290.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│        coq-fiat-crypto │ 3786.32 3796.34 -0.26 % │    10516481159588    10539476614002 -0.22 % │    16527919785618    16530098322822 -0.01 % │    2334888    2336228 -0.06 % │    1072     978    +9.61 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│        coq-lambda-rust │ 1526.89 1529.98 -0.20 % │     4255945102481     4262139709277 -0.15 % │     5646700255264     5637792099288 +0.16 % │    1477424    1481060 -0.25 % │       0       0     +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│            coq-unimath │ 3073.79 3079.25 -0.18 % │     8562257962815     8579532086053 -0.20 % │    15404796362866    15407973302118 -0.02 % │    3761044    3774284 -0.35 % │     596     545    +9.36 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│         coq-verdi-raft │ 1316.64 1318.68 -0.15 % │     3672397520260     3676508243756 -0.11 % │     4987729015889     4987340684401 +0.01 % │    1811308    1811348 -0.00 % │       0     105  -100.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│ coq-mathcomp-character │  181.66  181.94 -0.15 % │      505997325697      507255845935 -0.25 % │      669759252350      669793966554 -0.01 % │     845092     847148 -0.24 % │      99      43  +130.23 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│              coq-color │  696.78  697.00 -0.03 % │     1946091235560     1945294823675 +0.04 % │     2327521860977     2327162211101 +0.02 % │    1339852    1339892 -0.00 % │     306     503   -39.17 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│            coq-bignums │   68.96   68.97 -0.01 % │      191242708676      192014270524 -0.40 % │      253786520957      253769648570 +0.01 % │     491880     491696 +0.04 % │     183     156   +17.31 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│   coq-mathcomp-algebra │  165.37  165.16 +0.13 % │      459576132797      459995051268 -0.09 % │      568845304033      568839510083 +0.00 % │     609352     612216 -0.47 % │      16      14   +14.29 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│          coq-fiat-core │  108.10  107.92 +0.17 % │      307724213915      307527350654 +0.06 % │      385569641074      384981576798 +0.15 % │     484204     484192 +0.00 % │     642     242  +165.29 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│         coq-coquelicot │   75.88   75.72 +0.21 % │      207703072389      207988297992 -0.14 % │      251943295283      251596054142 +0.14 % │     653860     653968 -0.02 % │     480     204  +135.29 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│              coq-verdi │  119.20  118.72 +0.40 % │      331229536648      329968434377 +0.38 % │      418059547370      417521062464 +0.13 % │     558920     558420 +0.09 % │       8       6   +33.33 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼────────────────────────────┤
│             coq-sf-plf │   44.12   43.92 +0.46 % │      122673324601      122751929229 -0.06 % │      158931340360      158919558689 +0.01 % │     480676     488832 -1.67 % │     293      23 +1173.91 % │
└────────────────────────┴─────────────────────────┴─────────────────────────────────────────────┴─────────────────────────────────────────────┴───────────────────────────────┴────────────────────────────┘
~~~

**Review by SkySkimmer — APPROVED** [review]
Looking good

**Pull request merged**